crates/extensions/c8y_mapper_ext/src/tests.rs

```diff
@@ -2418,6 +2418,72 @@ async fn c8y_mapper_nested_child_service_event_mapping_to_smartrest() {
     .await;
 }
 
+#[tokio::test]
+async fn mapper_processes_operations_concurrently() {
+    let num_operations = 20;
+
+    let mut fts_server = mockito::Server::new();
+    let _mock = fts_server
+        .mock(
+            "GET",
+            "/tedge/file-transfer/test-device/config_snapshot/c8y-mapper-1234",
+        )
+        // make each download block so it doesn't complete before we submit all operations
+        .with_chunked_body(|_w| {
+            std::thread::sleep(Duration::from_secs(5));
+            Ok(())
+        })
+        .expect(num_operations)
+        .create_async()
+        .await;
+    let host_port = fts_server.host_with_port();
+
+    let cfg_dir = TempTedgeDir::new();
+    let (mqtt, http, _fs, _timer, _ul, dl) = spawn_c8y_mapper_actor(&cfg_dir, true).await;
+    let mut dl = dl.with_timeout(TEST_TIMEOUT_MS);
+
+    spawn_dummy_c8y_http_proxy(http);
+
+    let mut mqtt = mqtt.with_timeout(TEST_TIMEOUT_MS);
+    skip_init_messages(&mut mqtt).await;
+
+    // simulate many successful operations that needs to be handled by the mapper
+    for i in 0..num_operations {
+        mqtt.send(MqttMessage::new(
+            &Topic::new_unchecked(&format!("te/device/main///cmd/log_upload/c8y-mapper-{i}")),
+            json!({
+            "status": "successful",
+            "tedgeUrl": format!("http://{host_port}/tedge/file-transfer/test-device/log_upload/c8y-mapper-1234"),
+            "type": "mosquitto",
+            "dateFrom": "2023-11-28T16:33:50+01:00",
+            "dateTo": "2023-11-29T16:33:50+01:00",
+            "searchText": "ERROR",
+            "lines": 1000
+
+        })
+                .to_string(),
+        ))
+            .await.unwrap();
+
+        mqtt.send(MqttMessage::new(
+            &Topic::new_unchecked(&format!("te/device/main///cmd/config_snapshot/c8y-mapper-{i}")),
+            json!({
+            "status": "successful",
+            "tedgeUrl": format!("http://{host_port}/tedge/file-transfer/test-device/config_snapshot/c8y-mapper-1234"),
+            "type": "path/type/A",
+        })
+                .to_string(),
+        ))
+            .await.unwrap();
+    }
+
+    for _ in 0..(num_operations * 2) {
+        dl.recv()
+            .await
+            .expect("there should be one DownloadRequest per operation");
+    }
+}
+
```

Comment:
To validate the effectiveness of this work, one critical test that we really need to add, is to validate that a new request is still processed and mapped while a file upload, in response to a terminal status update of another client, is in-progress.

Reply:
will be done in a follow-up: https://github.com/thin-edge/thin-edge.io/issues/2880#issuecomment-2223213567